# SECRID  DATA ANALYSIS PROJECT


<span style="color: gray; font-size:1em;">March-2020</span>


## Table of Contents
* [Introduction](#introduction)
* [Section One - Import Data into IDE](#import_data)
    * [Part I - Gathering Data](#gather_data)
    * [Part II - Assessing Data](#assess_data)
    * [Part III - Cleaning Data](#clean_data)
* [Section Two - Individual Variable Assessing](#asses_variable)
    * [Variable 1 - internal_id](#Internal_ID)
    * [variable 2 - document_number](#Document_Number)
    * [variable 3 - customer_name](#Customer_Name)
        * [Quality Issue](#qi_Customer_Name)
    * [variable 4 - customer_category](#Customer_Category)
    * [variable 5 - retailer_role](#Retailer_Role)
        * [Quality Issue](#qi_Retailer_Role)
    * [variable 6 - pim_category](#PIM_Category)
    * [variable 7 - pim_colour](#PIM_Colour)
    * [variable 8 - item](#Item)
    * [variable 9 - display_name](#Display_Name)
    * [variable 10 - type_of_material](#Type_of_leather)
    * [variable 11 - wsl](#WSL)
    * [variable 12 - while_stock_lasts](#While_stock_lasts)
    * [variable 13 - shipping_country](#shipping_country)
        * [Quality Issue](#Customer_Name)
    * [variable 14 - date](#date)    
    * [variable 15 - quantity](#quantity)
    * [variable 16 - amount](#amount)
        * [Explore: quality and amount columns](#explore)
    
   
    
    
    
    
          
    

<a id='introduction'></a>
## Introduction

SECRID is a business entity based in the Netherlands. It produces, stocks and sells designer wallets, particularly leather based wallets in more than 100 countries all over the world.

This notebook explores SECRID sales data


<a id='#import_data'></a>
## Section One : Import Data into IDE

<a id='gather_data'></a>
## Part I : Gathering Data

In [1]:
# load required libraries
import numpy as np
import pandas as pd
from pandas import DataFrame
from pandas import read_excel
import xlsxwriter

import seaborn as sns
# Use seaborn style defaults and set the default figure size
sns.set(rc={'figure.figsize':(11, 4)})

import matplotlib.pyplot as plt
%matplotlib inline
import matplotlib.ticker as ticker
from matplotlib import pyplot as plt
from matplotlib import dates as mpl_dates
from pandas.plotting import register_matplotlib_converters
register_matplotlib_converters()
plt.style.use('seaborn')

import six
from datetime import datetime, timedelta

# environment settings:
pd.options.display.float_format = '{:.2f}'.format
pd.set_option('display.max_column',None)
pd.set_option('display.max_rows',None)
pd.set_option('display.max_seq_items',None)
pd.set_option('display.max_colwidth', 500)
pd.set_option('expand_frame_repr', True)


### Load .xlxs files

In [3]:
#load sales 2015 
df1 = pd.read_excel('SECRID_DATA.xlsx',0) #load first spreadsheet of SECRID DATA.xlxs

In [4]:
# #load sales 2016 
df2 = pd.read_excel('SECRID_DATA.xlsx',1) #load second spreadsheet of SECRID DATA.xlxs

In [5]:
#load sales 2017
df3 = pd.read_excel('SECRID_DATA.xlsx',2)  #load third spreadsheet ofSECRID DATA.xlxs

In [6]:
#load sales 2018
df4 = pd.read_excel('SECRID_DATA.xlsx',3)  #load fourth spreadsheet of SECRID DATA.xlxs

In [7]:
#load sales 2019
df5 = pd.read_excel('SECRID_DATA.xlsx',4)  #load fifth spreadsheet of SECRID DATA.xlxs

In [9]:
#combine df1, df2, df3, df4 and df5  into one complete dataframe 'df' for sales data
df = pd.concat([df1, df2, df3, df4, df5]) 

<a id='assess_data'></a>
## Part II - Assessing  Data

In [10]:
df.head() #preview first five rows

Internal ID Document Number       Date        Date Created  \
0        24560       I-1510000 2015-01-05 2015-01-05 12:35:00   
1        24580       I-1510001 2015-01-05 2015-01-05 16:15:00   
2        24580       I-1510001 2015-01-05 2015-01-05 16:15:00   
3        24580       I-1510001 2015-01-05 2015-01-05 16:15:00   
4        24580       I-1510001 2015-01-05 2015-01-05 16:15:00   

                                              Name  \
0                                 C-9855 SkilledIn   
1  C-9255 Rubino di Fabio Della Rocca attn. Spimar   
2  C-9255 Rubino di Fabio Della Rocca attn. Spimar   
3  C-9255 Rubino di Fabio Della Rocca attn. Spimar   
4  C-9255 Rubino di Fabio Della Rocca attn. Spimar   

              Customer Category Retailer Role     Shipping Address 1  \
0  Promotional sales - End user           NaN  Scheepsbouwweg 8, K11   
1                 Leather goods           NaN   Via Della Croce 75 A   
2                 Leather goods           NaN   Via Della Croce 75 A   
3                 Leather goods           NaN   Via Della Croce 75 A   
4                 Leather goods           NaN   Via Della Croce 75 A   

  Shipping Address 2 Shipping City Shipping Zip Shipping State/Province  \
0                NaN     Rotterdam      3089 JW                     NaN   
1                NaN          Roma        00187                     NaN   
2                NaN          Roma        00187                     NaN   
3                NaN          Roma        00187                     NaN   
4                NaN          Roma        00187                     NaN   

  Shipping Country         Item             Display Name  Quantity  Amount  \
0      Netherlands     TA-Brown  Twinwallet Amazon Brown         1   54.26   
1            Italy      C-Black      Cardprotector Black         1   11.20   
2            Italy       C-Blue       Cardprotector Blue         2   22.40   
3            Italy  C-Champagne  Cardprotector Champagne         1   11.20   
4            Italy        C-Red        Cardprotector Red         2   22.40   

   Amount (Foreign Currency)   PIM Category   Material PIM Colour WSL +  \
0                      54.26     Twinwallet     Amazon      Brown    No   
1                      11.20  Cardprotector  Aluminium      Black    No   
2                      22.40  Cardprotector  Aluminium       Blue    No   
3                      11.20            NaN        NaN        NaN    No   
4                      22.40  Cardprotector  Aluminium        Red    No   

  While stock lasts Cardprotector Colour  
0                No               Silver  
1                No                Black  
2                No                 Blue  
3                No                  NaN  
4                No                  Red

In [ ]:
df.tail() #preview last five rows

In [11]:
# Check size of the dataframe 
df.shape 

(2021470, 24)

In [ ]:
# list names of columns in dataframe
df.columns 

In [ ]:
# View info of the dataframe 
df.info()

In [ ]:
# view some of the core statistics about columns
df.describe(include='all')

In [ ]:
# check the Data types (dtypes) of each column in Dataframe
df.dtypes 

In [ ]:
# Total sum of duplicate rows
df.duplicated().sum() # returns a Boolean Series with True value for each duplicated row and sums them

In [ ]:
#return the number of unique elements in each column
print(df.nunique()) 

In [ ]:
df.count() #returns the number of non-missing values for each column or row

In [ ]:
#Total missing values(NaN) in a DataFrame
df.isnull().sum().sum()

In [ ]:
#Count number of NaN for each column in DataFrame
print(df.isnull().sum()) 

<a id='issues'></a>
**Quality issues**
 * Rename column names to have clear, descriptive names in small letters according to best practice. Column 'name' can be renamed to 'customer_name' and column 'material' can be renamed to 'type_of_material'
 * Set to columns to appropriate category data type: 'internal_id', 'document_number', 'customer_name', 'customer _category', 'retailer_role', 'shipping_country', 'item', 'display_name', 'pim_category','type_of_material', 
   'pim_colour', 'wsl', 'while_stock_lasts' and 'cardprotector_colour' 

<a id='clean_data'></a>
## Part III - Cleaning Data

In [12]:
# Create copy of original DataFrame
df_clean = df.copy()

In [13]:
#Fixing messy column names
df_clean.columns = df_clean.columns.str.strip().str.lower().str.replace(' ', '_').str.replace('(', '').str.replace(')', '')

In [14]:
# change column names using rename function
df_clean.rename(columns={                                                 
                         'name':'customer_name',
                         'wsl_+':'wsl',
                         'material':'type_of_material' }, 
                 inplace=True)

**Test**

In [15]:
df_clean.columns #List of column names in df_clean Dataframe

Index(['internal_id', 'document_number', 'date', 'date_created',
       'customer_name', 'customer_category', 'retailer_role',
       'shipping_address_1', 'shipping_address_2', 'shipping_city',
       'shipping_zip', 'shipping_state/province', 'shipping_country', 'item',
       'display_name', 'quantity', 'amount', 'amount_foreign_currency',
       'pim_category', 'type_of_material', 'pim_colour', 'wsl',
       'while_stock_lasts', 'cardprotector_colour'],
      dtype='object')

**Define**
<br>Set appropriate data types for fields mentioned in the [Quality issues](#issues) 

In [16]:
# use .astype to change data type of dataframe columns
df_clean = df_clean.astype({"internal_id":'category',"document_number":'category', "customer_name":'category', "customer_category":'category', "retailer_role":'category', "shipping_country":'category', "item":'category',"display_name":'category', "pim_category":'category', "type_of_material":'category', "pim_colour":'category',"wsl":'category', "while_stock_lasts":'category', "cardprotector_colour":'category'})

**Test**

In [ ]:
df_clean.info()

In [ ]:
# view some of the core statistics about columns
df_clean.describe(include='all')

### content structure of sales dataset
The sales data contains 24 columns (variables) and 2,021,470 rows (entries). 
This is evidence that 1,609,533 sale transactions were completed in the January 2015 – July 2019 period for SECRID business. The dataset contained features about:

* Products for sale: item, display_name, pim_category, pim_colour, type_of_material and cardprotector_colour
* The country the item was shipped to : shipping_country  
* Customer data:  customer_name, customer_category and retailer_role
* Sale transactions: internal_id, document_number, quantity, amount, amount(foreign_currency) and date


### Detected Missing Values
A null value is a value in a field that appears to be blank. A null value is a field with no value. 
The table below indicates the number and  the resulting percentage of missing values per column.This able is manually created with results from running; **df.count()** to get Value Count of non-missing values in a column and 
**print(df.isnull().sum())** which will give the count number of NaN for each column in DataFrame.

| Variable Name  | Value Count| Number Of Missing Values| % Of Missing Values
| -------------  | ------------- |------------- |-------------
| internal_id    | 2,021,470  |0 |0%
| document_number| 2,021,470  |0 |0%
| date           | 2,021,470  |0 |0%
| date_created   | 2,021,470  |0 |0%
| customer_name  | 2,021,470  |0 |0%
| customer_category  | 2,015,375  |6,095     |0.30%
| retailer_role      | 201,788    |1,819,682 |90.02%
| shipping_address_1 | 2,011,883  |9,587     |0.47%
| shipping_address_2 | 367,598    |1,653,872 |81.82%
| shipping_city      | 2,012,657  |8,813 |0.44%
| shipping_zip       | 1,990,236  |31,234|1.55%
| shipping_state/province |395,792 |1,625,678 |80.42%
| shipping_country   |2,014,891   |6,579 |0.33%
| item               | 2,021,470  |0 |0%
| display_name       | 2,021,470  |0 |0%
| quantity           | 2,021,470  |0 |0%
| amount             | 2,021,470  |0 |0%
| amount_foreign_currency | 2,021,470 |0 |0%
| type_of_material        | 1,941,770 |79,700 |3.94%
| pim_category            | 1,944,252 |77,218 |3.82%
| pim_colour              | 1,933,547 |87,923|4.35%
| wsl_+                   | 2,021,470 |0     |0%
| while_stock_lasts       | 2,021,470 |0     |0%
| cardprotector_colour    | 1,877,516 |143,954 |7.12%




<a id='asses_variable'></a>
## Section Two: Individual Variable assesment

<a id='Internal_ID'></a>
### VARIABLE 1  : internal_id

In [ ]:
df_clean.internal_id.describe()#overview of variable; count, unique, top,freq

In [ ]:
print (df_clean.internal_id.cat.categories)

In [ ]:
df_clean.internal_id.value_counts()  #Return  counts of unique values

In [ ]:
print(df_clean.internal_id.isnull().sum()) #Number of missing values in Internal_ID column

### Findings
This is a column of sale transaction identifiers.
'internal_id' column has 143,904  unique categories and no missing values.

<a id='Document_Number'></a>
### VARIABLE  2: document_number

In [ ]:
df_clean.document_number.describe()#overview of variable; count, unique, top,freq

In [ ]:
print (df_clean.document_number.cat.categories)#categories in the categorical variable

In [ ]:
df_clean.document_number.value_counts()  #Return  counts of unique values

In [ ]:
print(df_clean.document_number.isnull().sum())  #Number of missing values in Document_Number column

### Findings
This is a column of sale transaction identifiers.
'Document Number' column has 143,890  unique categories and no missing values.

<a id='Customer_Name'></a>
### VARIABLE  3: customer_name

In [ ]:
df_clean.customer_name.describe()#overview of variable; count, unique, top,freq

In [ ]:
print (df_clean.customer_name.cat.categories)#categories in the categorical variable

In [ ]:
df_clean.customer_name.value_counts()  #Return counts of unique values

In [ ]:
print(df_clean.customer_name.isnull().sum())  #Number of missing values in Customer_Name column

### Findings
This column records the customer name per sale transaction. The format the Customer Name is entered is as follows; the letter C (capital c) is entered as a prefix before a hyphen (‘-’), a unique number is then added followed by the customer name. There are 31,143 unique customer names. This means that for the January 2015–July 2019 period the business  had 31,143 unique customers. The customer who performed the most transactions is **C-10625 El Corte Ingles** at 35,481 sale transactions.

<a id='qi_Customer_Name'></a>
**Quality Issue**

For customers with different branches of their business, each branch is considered as a unique customer.
The letter C (capital c) that is entered as a prefix before a hyphen (‘-’) and the unique number following is repeated for all branches.<br>
For example;

* 'C-10215 James Shoe Care Canada Place Limited',
* 'C-10215| C-1 James Shoe Care Canada Place Limited : James Shoe Care Canada Place Limited : James Shoe Care Wharf',
* 'C-10215| C-2 James Shoe Care Canada Place Limited : James Shoe Care Canada Place Limited : James Shoe Care Westfield',


### combine similar branch stores sale transactions.

In [ ]:
#create new column that has first 7 characters of column 'customer_name' 
df_clean['customer_id'] = df_clean['customer_name'].str[:7]

**Test**

In [ ]:
df_clean.head()

In [ ]:
df_clean = df_clean.astype({"customer_id":'category'})

In [ ]:
df_clean.customer_id.describe()

In [ ]:
df_clean.customer_id.value_counts()

In [ ]:
df_clean.customer_id= df_clean.customer_id.astype('category')

In [ ]:
df_clean.customer_id.cat.categories#categories in the categorical variable

In [ ]:
df_clean.loc[df_clean['customer_id'] == 'C-6962|'].head()

In [ ]:
df_clean.loc[df_clean['customer_id'] == 'C-6962'].head()

In [ ]:
df_clean.customer_id = df_clean.customer_id.str.strip().str.replace('|', '')

**Test**

In [ ]:
df_clean.customer_id.describe()

In [ ]:
df_clean.customer_id.value_counts()

<a id='Customer_Category'></a>
### VARIABLE  4: customer_category

In [ ]:
df_clean.customer_category.describe()#overview of variable; count, unique, top,freq

In [ ]:
print (df_clean.customer_category.cat.categories)#categories in the categorical variable

In [ ]:
df_clean.customer_category.value_counts() #Return counts of unique values

In [ ]:
print(df_clean.customer_category.isnull().sum())  #Number of missing values in Customer_Name column

### Findings

'Customer Category' column has 47  unique categories and 5,415 missing values. Leather goods category has the highest frequency of sale transactions at 345,551.

<a id='Retailer_Role'></a>
### VARIABLE  5: retailer_role

In [ ]:
df_clean.retailer_role.describe()#overview of variable; count, unique, top,freq

In [ ]:
print (df_clean.retailer_role.cat.categories)#categories in the categorical variable

In [ ]:
df_clean.retailer_role.value_counts()#Return counts of unique values

In [ ]:
print(df_clean.retailer_role.isnull().sum()) #Number of missing values in Retailer_Role column

### Findings

'Retailer Role' column has 4 unique categories and 1,563,539 missing values. This column has 97% missing values and is recommended to be dropped during cleaning.

<a id='qi_Retailer_Role'></a>
### Quality Issues
* 'Retailer Role' column : This column has 1,563,539 missing values. This column has 97% missing values and is recommended to be dropped

In [ ]:
# Drop 'retailer_role' column with drop function
df_clean.drop(['retailer_role'], axis=1)

<a id='PIM_Category'></a>
### VARIABLE 6 : pim_category

In [ ]:
df_clean.pim_category.describe() #overview of variable; count, unique, top,freq

In [ ]:
# Main categories of items sold
# print categories in the categorical variable
print (df_clean.pim_category.cat.categories)

In [ ]:
df_clean.pim_category.value_counts() #Return counts of unique values

In [ ]:
print(df_clean.pim_category.isnull().sum()) 

### Findings
Over the January 2015–July 2019 period the product sold by the business are divided into six (6) distinct main categories.
* Cardprotector
* Cardslide
* Miniwallet
* Moneyband
* Slimwallet
* Twinwallet

'PIM_Category' column has 6 unique categories and 1,040 missing values. **Miniwallet** category has the highest frequency of sale transactions at 901,668.

<a id='PIM_Colour'></a>
### VARIABLE  7: pim_colour

In [ ]:
df_clean.pim_colour.describe() #overview of variable; count, unique, top,freq

In [ ]:
# Main categories in PIM Colour
print (df_clean.pim_colour.cat.categories)

In [ ]:
df_clean.pim_colour.value_counts() #Return counts of unique values

In [ ]:
print(df_clean.pim_colour.isnull().sum()) 

### Findings

This column indicates color of the product sold.'PIM Colour' column has 45 unique categories and 6,058 missing values. **Black** category has the highest frequency of sale transactions at 465,688.


<a id='Item'></a>
### VARIABLE 8: item

In [ ]:
df_clean.item.describe() #overview of variable; count, unique, top,freq

In [ ]:
# print categories in the categorical variable
print (df_clean.item.cat.categories)

In [ ]:
df_clean.item.value_counts() #Return counts of unique values

In [ ]:
print(df_clean.item.isnull().sum()) 

### Findings

Once the items are classified into a main category (PIM Category), they are later categorized into more detailed sub-categories based on the type of material the item is made of (Type of Material) and product color (PIM Colour). The initials of the PIM Category and Type of Material are combined as a prefix before a hyphen (‘-’) and the PIM Colour is added after the hyphen. 

'Item' column has 292 unique categories and no missing values. Unexpected numeric type categories are spotted.This type of entries will technically be considered as a missing values. **M-Black** category has the highest frequency of sale transactions at 65,720.

<a id='Display_Name'></a>
### VARIABLE  9: display_name

In [ ]:
df_clean.display_name.describe() #overview of variable; count, unique, top,freq

In [ ]:
print (df_clean.display_name.unique()) 

In [ ]:
print (df_clean.display_name.cat.categories) #categories in the categorical variable

In [ ]:
df_clean.display_name.value_counts()#Return counts of unique values

In [ ]:
print(df_clean.display_name.isnull().sum()) #Number of missing values in Display_Name column

### Findings
'Display Name' like the Item column, gives a detailed description of the item sold. This column however, gives the full product name unlike the Item column entries. 'Display Name' column has 288 unique categories and no missing values. **Miniwallet Original Black** category has the highest frequency of sale transactions at 65,720.

<a id='Type_of_leather'></a>
### VARIABLE 10 : type_of_material

In [17]:
df_clean.type_of_material.describe() #overview of variable; count, unique, top,freq

count     1941770
unique         29
top       Vintage
freq       569304
Name: type_of_material, dtype: object

In [18]:
#Types of leather used in production
print (df_clean.type_of_material.cat.categories) # Get list of categories in categorical variable

Index(['Aluminium', 'Amazon', 'Cleo', 'Crisple', 'Cubic', 'Dash', 'Diamond',
       'Dutch Martin', 'Elastic', 'Engraved Aluminium', 'Glamour', 'Indigo',
       'Matte', 'Metallic', 'Nile', 'Optical', 'Original', 'Ornament',
       'Perforated', 'Polished', 'Polycarbonate', 'Powder coated Aluminium',
       'Prism', 'Rango', 'Recycled', 'Vegan', 'Vegetable Tanned',
       'Vegetable Tanned Stitched', 'Vintage'],
      dtype='object')


In [19]:
df_clean.type_of_material.value_counts() #count per category

Vintage                      569304
Original                     301510
Aluminium                    273787
Matte                        159531
Crisple                      124691
Dutch Martin                  72896
Rango                         64007
Elastic                       51398
Nile                          45831
Polycarbonate                 33187
Recycled                      29897
Cubic                         29678
Metallic                      26825
Indigo                        25513
Vegetable Tanned              23342
Engraved Aluminium            22600
Prism                         21965
Vegetable Tanned Stitched     19730
Amazon                        12518
Ornament                       6829
Cleo                           4802
Optical                        4679
Diamond                        4392
Dash                           4373
Glamour                        3511
Vegan                          3360
Perforated                     1388
Polished                    

In [20]:
print(df_clean.type_of_material.isnull().sum()) 

79700


### Findings
This column indicates the type of material used in creating the product.
The business uses 29 unique types of material. This column has 1,040 missing values.It is noted that missing values in this column imply that the item is not made of leather.  **Vintage** category has the highest frequency of sale transactions at 483,982.

<a id='WSL'></a>
### VARIABLE 11 : wsl

In [ ]:
df_clean.wsl.describe() #overview of variable; count, unique, top,freq

In [ ]:
# Get list of categories in categorical variable
print (df_clean.wsl.unique())

#or 

print (df_clean.wsl.cat.categories)

In [ ]:
df_clean.wsl.value_counts() #Return counts of unique values

In [ ]:
print(df_clean.wsl.isnull().sum()) 

### Findings
This column enters either 'Yes' or 'No' if an Item is currently running out of stock and is no longer produced.
This column has one constant entry which is ‘No’. This means that no item is currently running out of stock and is no longer being produced.
It is recommended that the column be dropped during data cleaning.This column has no missing values.

<a id='While_stock_lasts'></a>
### VARIABLE 12 : while_stock_lasts

In [ ]:
df_clean.while_stock_lasts.describe()  #overview of variable; count, unique, top,freq

In [ ]:
# Get list of categories in categorical variable
print (df_clean.while_stock_lasts.cat.categories)

In [ ]:
df_clean.while_stock_lasts.value_counts() #Return counts of unique values

In [ ]:
print(df_clean.while_stock_lasts.isnull().sum()) 

### Findings
This column enters either yes or no if an Item has been taken out of the collection, but the business still has the leather in stock at the business suppliers.  **No** category has the highest frequency of sale transactions at 1,378,556. 

<a id='shipping_country'></a>
### VARIABLE  13: shipping_country

In [ ]:
df_clean.shipping_country.describe()#overview of variable; count, unique, top,freq

In [ ]:
df_clean = df_clean.astype({"shipping_country":'category'})

In [ ]:
# List of countries the business shipped items to
print (df_clean.shipping_country.cat.categories)# Get list of categories in categorical variable

In [ ]:
# Total Sales transactions per country
df_clean.shipping_country.value_counts() #count per category(Transactions per country)

In [ ]:
print(df_clean.shipping_country.isnull().sum()) 

### Findings
For every sale transaction, the country the item was shipped to is indicated in this column. 
This variable has 102 unique categories implying that for the January 2015–July 2019 period the business shipped items to 102 different countries. 
This column has 6,102 missing values.  **Netherlands** category has the highest frequency of sale transactions at 371,805. 

### Quality Issues
* 'shipping_country' column : rename country category names to the appropriate official country names.

In [ ]:
# fix messy category names
df_clean.shipping_country = df_clean.shipping_country.str.strip().str.replace(',', '').str.replace("'", '')

In [ ]:
#rename category names using .cat.rename_categories
df_clean.shipping_country.cat.rename_categories({"Croatia/Hrvatska":'Republic of Croatia',"Iran (Islamic Republic of)":'Iran',"Vanuatu":'Republic of Vanuatu',"Korea Democratic Peoples Republic":"Democratic People's Republic of Korea","Korea Republic of":'Republic of Korea'})

<a id='date'></a>
### VARIABLE 14 : Date

In [ ]:
df_clean['year'] = df_clean.date.dt.year # create new column 'year' that registers year sale transaction was held.(helps with analysis)

**Test**

In [ ]:
df_clean.head()

In [ ]:
df_clean.year.value_counts().sort_index() # number of sale transactions per year

<a id='quantity'></a>
### VARIABLE 15 : quantity

In [ ]:
# Maximum Quantity of items sold in a trasaction for 2015-2019
df_clean.quantity.max()

In [ ]:
# Minimum Quantity of items sold per day for 2015-2019
df_clean.quantity.min()

In [ ]:
# Plot the data
plt.rcParams['figure.figsize']=(15,10)
plt.plot_date(df_clean.index, df_clean.quantity,linestyle='solid')
# Add a legend
plt.legend(['Quantity of items sold per transaction'],fontsize=18)
plt.xlabel('Date', fontsize=18)
plt.ylabel('Quantity',fontsize=18)
plt.title( 'QUANTITY OF ITEMS SOLD PER TRANSACTION OVER JAN,2015-JULY,2019',fontsize=20,y=1.03)
plt.tick_params(labelsize=12)
plt.tight_layout()
plt.savefig('qty.png',type="png",dpi=300)#Export graph as .png
# Show the plot
plt.show()

### Findings
From the figure above we note that there are negative Quantity entries. 
Quantity column indicates the number of items purchased in a transaction, it is therefore unexpected to find negative entries.

The negative entries in the Quantity column are records of items that have been returned by the customer after purchase for one reason or another.

The **overall totals and means will be brought down** if Quality data is assesed or analyzed as is. We shall therefore analyze the returned items (indicated as -ve Quantity) and those purchased but not returned (indicated as +ve Quantity) separately.


### Quantity of Purchased items(not returned)

In [ ]:
#Total quantity of items sold for 2015-2019 (not returned)
sum(qty for qty in df_clean.quantity if qty > 0) #total sum of '+ve' quantity

### Quantity of returned items upon purchase 

In [ ]:
# Dataframe of returned items
qtynegative = df_clean[(df_clean['quantity']<0)]# Rows with negative values in Quantity column
qtynegative.head()

In [ ]:
qtynegative.info()

In [ ]:
qtynegative.to_csv('qtynegative.csv', index=False)

In [ ]:
# sale transactions where items were returned
sum(qty < 0 for qty in df_clean.quantity)# Negative(-ve) entries in Quantity column

In [ ]:
# Total Quantity of items that were returned by customers after purchase
sum(qty for qty in df_clean.quantity if qty < 0)#sum the total (-ve) items 

## Findings
The maximum quantity of items sold in a sale transaction for January 2015 – July 2019 period was 2,510 while the minimum was -1,000. 
Excluding the returned items upon purchase (negative quantity entries), the total quantity of products/ items sold over the January 2015 – July 2019 period is 5,548,484. 


<a id='amount'></a>
### VARIABLE 16 : amount

In [ ]:
# Maximum revenue generated in a sale trasaction for 2015-2019
df_clean['amount'].max()

In [ ]:
# Minimum revenue generated in a sale trasaction for 2015-2019
df_clean['amount'].min()

In [ ]:
# Plot the data
plt.rcParams['figure.figsize']=(15,10)
plt.plot_date(df_clean.index, df_clean.amount,linestyle='solid')
# Add a legend
plt.legend(['Revenue generated per transaction'], fontsize=18)
plt.xlabel('Date', fontsize=18)
plt.ylabel('Revenue Amount(EUR)', fontsize=18)
plt.title( 'AMOUNT GENERATED PER TRANSACTION OVER JAN 2015-JULY 2019', fontsize=20)
plt.tick_params(labelsize=18)
plt.tight_layout()
plt.savefig('amt.png',type="png",dpi=300)#Export graph as .png
# Show the plot
plt.show()

In [ ]:
# Dataframe of returned items
amtpositive = df_clean[(df_clean['amount']>0)]# Rows with negative values in Quantity column
amtpositive.head()

In [ ]:
amtpositive.info()

In [ ]:
amtpositive.to_csv('amtpositive.csv', index=False)

In [ ]:
#Total revenue generated for January 2015-2019
sum(amount for amount in df_clean.amount if amount > 0) #total sum of '+ve' quantity

In [ ]:
#Sale transactions of reimbursed amount
sum(amount < 0 for amount in df_clean.amount)# Negative entries in amount column

In [ ]:
#sum total of reimbursed amount (revenue lost)
sum(amount for amount in df_clean.amount if amount < 0)

<a id='explore'></a>
###  Explore: quantity and amount columns

In [ ]:
# rows with positive values in Quantity column and Amount column
dfrevenue = df_clean[(df_clean.amount>0) & (df_clean.quantity>0)]

In [ ]:
dfrevenue.head()

In [ ]:
dfrevenue.to_csv('dfrevenue.csv', index=False)

In [ ]:
dfrevenue.info()

In [ ]:
dfrevenue.amount.max() #maximum revenue generated in a transaction

In [ ]:
dfrevenue.amount.min() #minimum revenue generated in a transaction

In [ ]:
dfrevenue.amount.mean() #average qamount of items bought per transaction

In [ ]:
dfrevenue.amount.mode()

In [ ]:
# Group the data frame by year and extract a number of stats per year
revenuerank = dfrevenue.groupby(
   ['year']
).agg(
    {
         'amount':sum,    # Sum revenue per customer
         'quantity': sum  # get the sum of items sold per year
         
    }
)

revenuerank.sort_values(by=['year'], inplace=True, ascending=True)

In [ ]:
revenuerank.head()

In [ ]:
# rows with negative values in Quantity column and Amount column
dfreimbursed  = df_clean[(df_clean.amount<0) & (df_clean.quantity<0)]

In [ ]:
dfreimbursed.head()

In [ ]:
dfreimbursed.to_csv('dfreimbursed.csv', index=False)

In [ ]:
dfreimbursed.info()

In [ ]:
reimbursedrank = dfreimbursed.groupby(
   ['year']
).agg(
    {
         'amount':sum,    # Sum revenue per customer
         'quantity': sum  # get the sum of items sold per year
         
    }
)

reimbursedrank.sort_values(by=['year'], inplace=True, ascending=True)

In [ ]:
reimbursedrank.head()

In [ ]:
#zero quantity and zero revenue rows (9 rows)
zerorevenue = df_clean[(df_clean.amount == 0) & (df_clean.quantity == 0)]

In [ ]:
zerorevenue.to_csv('zerorevenue.csv', index=False)

In [ ]:
zerorevenue.head(9)

In [ ]:
zerorevenue.info()

In [ ]:
zerorevenuerank = zerorevenue.groupby(
   ['year']
).agg(
    {
         'amount':sum,    
         'quantity': sum  
         
    }
)

zerorevenuerank.sort_index(inplace=True)

In [ ]:
zerorevenuerank

In [ ]:
#Negative quantity and zero revenue (or +ve) (749 rows)
notreimbursed = df_clean[(df_clean.amount >= 0) & (df_clean.quantity < 0)]

In [ ]:
notreimbursed.head(5)

In [ ]:
notreimbursed.to_csv('notreimbursed.csv', index=False)

In [ ]:
notreimbursed.info()

In [ ]:
notreimbursedrank = notreimbursed.groupby(
   ['year']
).agg(
    {
         'amount':sum,    # Sum revenue per customer
         'quantity': sum  # get the sum of items sold per year
         
    }
)
notreimbursedrank.sort_index(inplace=True)

In [ ]:
notreimbursedrank

In [ ]:
#zero (or -ve) revenue rows but +ve quantity (unpaid goods)(8,356 rows)
unpaidrevenue = df_clean[(df_clean.amount <= 0) & (df_clean.quantity > 0)]

In [ ]:
unpaidrevenue.head()

In [ ]:
unpaidrevenue.to_csv('unpaidrevenue.csv', index=False)

In [ ]:
unpaidrevenue.tail()

In [ ]:
unpaidrevenue.info()

In [ ]:
unpaidrevenuerank = unpaidrevenue.groupby(
   ['year']
).agg(
    {
         'amount':sum,    # Sum revenue per customer
         'quantity': sum  # get the sum of items sold per year
         
    }
)

unpaidrevenuerank.sort_index(inplace=True)

In [ ]:
unpaidrevenuerank

## Quality Issues

* 'Item' column : Unexpected numeric type categories are spotted.This type of entries will technically be considered as a missing values.

* 'quantity' and 'amount' columns: 
    * Filter out zero quantity and zero revenue rows


## DATA CLEANING

## EXPLARATORY DATA ANALYSIS

In [ ]:
# rows with positive values in Quantity column and Amount column
dfrevenue = df_clean[(df_clean.amount>=9.95) & (df_clean.quantity>0)]

In [ ]:
netherlandsrevenue =dfrevenue[(dfrevenue.shipping_country == 'Netherlands')]

In [ ]:
netherlandsrevenue.info()

In [ ]:
netherlandsrevenue.amount.max() #maximum revenue generated in a transaction

In [ ]:
netherlandsrevenue.amount.min() 

In [ ]:
netherlandsrevenuerank = netherlandsrevenue.groupby(
   ['year']
).agg(
    {
         'amount':sum,    # Sum revenue per customer
         'quantity': sum  # get the sum of items sold per year
         
    }
)

netherlandsrevenuerank.sort_values(by=['year'], inplace=True, ascending=True)

In [ ]:
netherlandsrevenuerank

In [ ]:
customerrank = dfrevenue.groupby(
   ['year', 'customer_name', 'customer_category']
).agg(
    {
         'amount':sum,    # Sum revenue per customer
         'quantity': sum  # get the sum of items sold per year
         
    }
)

customerrank.sort_values(by=['amount'], inplace=True, ascending=False)

In [ ]:
customerrank

## TIME SERIES ANALYSIS

In [24]:
# CREATE TIME SERIES
df_clean.set_index('date', inplace=True) #set 'date' column as dataframe index

**Test**

In [ ]:
df_clean.index

## MINIWALLET

In [ ]:
miniwalletdf =df_clean[(df_clean.pim_category == 'Miniwallet')]

In [ ]:
miniwalletdf.head() #preview first five rows

In [ ]:
timelessdesign= miniwalletdf.groupby(
   ['type_of_material']
).agg(
    {
         'amount':sum,    # Sum revenue per customer
         'quantity': sum  # get the sum of items sold per year
         
    }
)

timelessdesign.sort_values(by=['amount'], inplace=True, ascending=False)